viewer3d/base.py

```diff
@@ -176,74 +199,219 @@ def reset(self):
+        assert isinstance(poses, list), "The 'poses' argument must be of type `list`."
+        for pose in poses:
+            assert isinstance(
+                pose, Pose
+            ), f"Object must be of type `Pose`. Received: {type(pose)}"
+        self.poses[index] = poses
+        self.pdbstrings[index] = [None] * len(poses)
+        if update_viewer:
+            self.update_viewer(index=index)
+
+    @requires_show
+    def update_pdbstrings(self, pdbstrings, index=None, update_viewer=True):
+        if index is None:
+            index = self.get_decoy_widget_index()
+        assert isinstance(
+            pdbstrings, list
+        ), "The 'pdbstrings' argument must be of type `list`."
+        for pdbstring in pdbstrings:
+            assert isinstance(
+                pdbstring, str
+            ), f"Object must be of type `str`. Received: {type(pdbstring)}"
+        self.poses[index] = [None] * len(pdbstrings)
+        self.pdbstrings[index] = pdbstrings
+        if update_viewer:
+            self.update_viewer(index=index)
+
+
 @attr.s(kw_only=False, slots=False)
 class WidgetsBase:
-    def set_widgets(self, obj):
-        self.widgets = _to_widgets(obj)
+    decoy_widget = attr.ib(
+        default=attr.Factory(
+            lambda self: interactive(
+                self.update_viewer,
+                index=IntSlider(
+                    min=0,
+                    max=self.n_decoys - 1,
+                    description="Decoys",
+                    continuous_update=self.continuous_update,
+                ),
+            ),
+            takes_self=True,
+        ),
+    )
+
+    def get_decoy_widget_index(self):
+        kwargs = self.decoy_widget.kwargs
+        index = kwargs["index"] if kwargs else 0
+        return index
 
     def get_widgets(self):
         _widgets = self.widgets.copy()
         if self.n_decoys > 1:
             _widgets.insert(0, self.decoy_widget)
+        else:
+            self.update_viewer()
         return _widgets
 
-    def get_decoy_widget(self):
-        return interactive(
-            self.update_decoy,
-            index=IntSlider(
-                min=0,
-                max=self.n_decoys - 1,
-                description="Decoys",
-                continuous_update=self.continuous_update,
-            ),
-        )
+    def get_widgets_dict(self):
+        return {
+            widget.description: widget.value
+            for widget in self.widgets
+            if all(hasattr(widget, attr) for attr in ("description", "value"))
+        }
+
+    def set_widgets(self, obj):
+        self.widgets = _to_widgets(obj)
+
+    def update_viewer(self, index: Optional[int] = None):
+        time.sleep(self.delay)
+        if index is None:
+            index = self.get_decoy_widget_index()
+        if index in self.poses.keys():
+            self.update_objects(self.poses[index], self.pdbstrings[index])
+
+    def _maybe_setup_colab(self):
```

Comment:
@ajasja `_maybe_setup_colab` runs upon instantiation of the viewer - hopefully this helps show the viewer in Google Colab!

Reply:
Thanks! Nglview now displays!